File: jar/src/main/java/jline/api/measures/Ms_neymanchisquared.java
package jline.api.measures;

import jline.util.matrix.Matrix;

public final class Ms_neymanchisquared {
    private Ms_neymanchisquared() {}

    /**
     * Neyman chi-squared distance between two probability distributions.
     * Part of the squared L2 or chi-squared family.
     *
     * @param P first probability distribution
     * @param Q second probability distribution
     * @return Neyman chi-squared distance
     */
    public static double ms_neymanchisquared(Matrix P, Matrix Q) {
        if (P.length() != Q.length()) {
            throw new IllegalArgumentException("Distributions must have the same size");
        }

        double sum = 0.0;
        for (int i = 0; i < P.length(); i++) {
            double pi = P.get(i);
            double qi = Q.get(i);
            if (pi > 0) {
                double diff = pi - qi;
                sum += (diff * diff) / pi;
            }
        }

        return sum;
    }
}

File: jar/src/main/kotlin/jline/api/measures/Ms_neymanchisquared.kt
package jline.api.measures

import jline.util.matrix.Matrix

/**
 * Neyman chi-squared distance between two probability distributions.
 * Part of the squared L2 or chi-squared family.
 * 
 * @param P first probability distribution
 * @param Q second probability distribution
 * @return Neyman chi-squared distance
 */
fun ms_neymanchisquared(P: Matrix, Q: Matrix): Double {
    require(P.length() == Q.length()) { "Distributions must have the same size" }
    
    var sum = 0.0
    for (i in 0 until P.length()) {
        val pi = P.get(i)
        val qi = Q.get(i)
        if (pi > 0) {
            val diff = pi - qi
            sum += (diff * diff) / pi
        }
    }
    
    return sum
}
/**
 * Neymanchisquared metric algorithms
 */
@Suppress("unused")
class MsNeymanchisquaredAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}